### ISSUE TITLE
Exception information is lost for async code

### ISSUE THREAD

**Issue by trampster** [xamarin/xamarin-android#7161]
### Android application type

Classic Xamarin.Android (MonoAndroid12.0, etc.)

### Affected platform version

VS2022

### Description

When an unhandled exception occurs in async code instead of the actual type and stack trace you get the following:

```
[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object.
[mono-rt]   at Android.Runtime.JNINativeWrapper._unhandled_exception (System.Exception e) [0x0000e] in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:12 
[mono-rt]   at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V (_JniMarshal_PP_V callback, System.IntPtr jnienv, System.IntPtr klazz) [0x0001c] in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:23 
[mono-rt]   at (wrapper native-to-managed) Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V(intptr,intptr)
```

It doesn't seem to matter what the original type was it's always shown as a NullReferenceException with that stack trace.

If your debugging then you get an Unhandled Exception popup that has no exception information except the wrong type (NullReferenceException) and no stack trace (it also doesn't stop on the location that threw the exception) if you then continue you get the [mono-rt] exception log from above.

This makes it almost impossible to find out/debug the problem as all relevant information is lost.

### Steps to Reproduce

1. Create a new Xamarin.Forms project
2. Put the following code into the MainActivity.OnCreate
```
     await Task.Delay(1000);
     throw new InvalidOperationException("This exception info should not be lost");
```
3. Run

### Did you find any workaround?

no

**Comment by trampster** [comment]
Here is my Visual Studio 2022 version info:

Microsoft Visual Studio Professional 2022
Version 17.2.5
VisualStudio.17.Release/17.2.5+32616.157
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

.NET Core Debugging with WSL   1.0
.NET Core Debugging with WSL

ASP.NET and Web Tools 2019   17.2.393.26812
ASP.NET and Web Tools 2019

Azure App Service Tools v3.0.0   17.2.393.26812
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   17.2.393.26812
Azure Functions and Web Jobs Tools

C# Tools   4.2.0-4.22281.5+8d3180e5f00d42f0f0295165f756f368f0cbfa44
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Mono Debugging for Visual Studio   17.2.20 (482eb2a)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   6.2.1
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Razor (ASP.NET Core)   17.0.0.2218101+885a343b00bcab620a90c1550c37dafd730ce984
Provides languages services for ASP.NET Core Razor.

SQL Server Data Tools   17.0.62204.01010
Microsoft SQL Server Data Tools

TypeScript Tools   17.0.10418.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.2.0-4.22281.5+8d3180e5f00d42f0f0295165f756f368f0cbfa44
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   17.1.0-beta.22274.4+211fd760c2315d73cfa0d375fe64abb346bc1096
Microsoft Visual F# Tools

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VSPackage Extension   1.0
VSPackage Visual Studio Extension Detailed Info

Xamarin   17.2.0.177 (d17-2@2f6c881)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   17.2.0.244 (remotes/origin/d17-2@197e1a0b7)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   17.2.15 (2e3b60e)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   12.3.3.3 (d17-2/4e061b7)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: dffa5ab
    Java.Interop: xamarin/java.interop/d17-2@9760f0a9
    ProGuard: Guardsquare/proguard/v7.0.1@912d149
    SQLite: xamarin/sqlite/3.38.2@7b1e016
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d17-2@fc3c2ac


Xamarin.iOS and Xamarin.Mac SDK   15.10.0.5 (96b3edb6d)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

**Comment by codedbycurtis** [comment]
Can confirm this is happening in my Xamarin.Forms app as well.

I think that due to the nature of async void methods, the exception object is swallowed, which is standard behaviour. However, for whatever reason, the Android runtime still attempts to propagate the swallowed exception. Because the exception has been swallowed, the exception object being propagated by the runtime is evaluated to `null`, which results in a `System.NullReferenceException` being thrown from the `JNINativeWrapper`.

**Comment by trampster** [comment]
async void exceptions are not swallowed, they are treated as unhandled exceptions and therefore should stop the app. (in general we avoid async void because you can't catch/handle them from further app the stack, not because they are swallowed)

**Comment by grendello** [comment]
This is most likely fixed by https://github.com/xamarin/xamarin-android/pull/7103, the fix will be part of the next VS update. Please test when the fix is available and reopen this issue if the problem persists. Thanks!

**Comment by trampster** [comment]
Just got an update to Version 17.2.6 and unfortunately this is not fixed.

I can't see how to reopen this issue.

**Comment by codedbycurtis** [comment]
The issue does seem to be fixed in Version 17.3 Preview 3, so I imagine we'll have to wait until 17.3 is available.

**Comment by grendello** [comment]
That's correct, the update will be part of the 17.3 release. I should have been more precise, apologies :)
